# Recommendation System

## Business Understanding

The analysis below is for a custom recommendation system for songs.  The data chosen for training/testing came from the [Million Song Dataset Challenge](https://www.kaggle.com/c/msdchallenge#description). The dataset was originally provided by The Echo Nest and made available by [Columbia University](https://labrosa.ee.columbia.edu/millionsong/) with the puprose encouraging research in machine learning.  The predictive modeling will be done to determine songs the user would like to listen to based on the current song playing.  

## Data Understanding

The analysis will involve two datasets.   The song _usage_ data and the song _information_ data.

The first is the Taste Profile subset.   The data includes a unique hash value for each user along with the song identification number and the number of times the user has played that particular song.  The data was not gathered over a specified time.  Rather the data is dependent upon the user having a minimum of 10 unique songs.  

The second dataset is the song data which provides details about the song.  The usage data only contains the song identification number while the song data provides the song title, artist, release, and year.  

### Downloading the usage and song data

In [1]:
%%capture

import graphlab as gl
import graphlab.aggregate as agg
from matplotlib import pyplot as plt

gl.canvas.set_target('ipynb')

usage_data = gl.SFrame.read_csv("../data/kaggle_visible_evaluation_triplets.txt",
                                header=False,
                                delimiter='\t',
                                column_type_hints={'X3':int})

songs = gl.SFrame.read_csv("../data/song_data.csv")

### Change data labels to be human readable

In [2]:
%%capture
print(usage_data.rename({'X1':'user', 'X2': 'song_id', 'X3': 'plays'}))

The usage data has three fields (`user`, `song_id`, and `plays`) which describes how many times (`play`) a `song` is listened to by a `user`.

In [3]:
usage_data.head(4)

Columns:
	user	str
	song_id	str
	plays	int

Rows: 4

Data:
+-------------------------------+--------------------+-------+
|              user             |      song_id       | plays |
+-------------------------------+--------------------+-------+
| fd50c4007b68a3737fe052d5a4... | SOBONKR12A58A7A7E0 |   1   |
| fd50c4007b68a3737fe052d5a4... | SOEGIYH12A6D4FC0E3 |   1   |
| fd50c4007b68a3737fe052d5a4... | SOFLJQZ12A6D4FADA6 |   1   |
| fd50c4007b68a3737fe052d5a4... | SOHTKMO12AB01843B0 |   1   |
+-------------------------------+--------------------+-------+
[4 rows x 3 columns]

The song data has five fields (`song_id`, `title`, `release`, `artist_name`, and `year`.

In [4]:
songs.head(4)

Columns:
	song_id	str
	title	str
	release	str
	artist_name	str
	year	int

Rows: 4

Data:
+--------------------+-------------------+-----------------------+
|      song_id       |       title       |        release        |
+--------------------+-------------------+-----------------------+
| SOQMMHC12AB0180CB8 |    Silent Night   | Monster Ballads X-Mas |
| SOVFVAK12A8C1350D9 |    Tanssi vaan    |      Karkuteillä      |
| SOGTUKN12AB017F4F1 | No One Could Ever |         Butter        |
| SOBNYVR12A8C13558C |   Si Vos Querés   |        De Culo        |
+--------------------+-------------------+-----------------------+
+------------------+------+
|   artist_name    | year |
+------------------+------+
| Faster Pussy cat | 2003 |
| Karkkiautomaatti | 1995 |
|  Hudson Mohawke  | 2006 |
|   Yerba Brava    | 2003 |
+------------------+------+
[4 rows x 5 columns]

To enhance the `usage_data` records details from `songs` data will be included.   The inclusion of the details will allow for subjective review as well as expand the analysis from song usage to artist usage.  A merge on the `usage_data` and `songs` datasets will provide the needed details.

In [6]:
ud_df = usage_data.to_dataframe()
song_df = songs.to_dataframe()

new_df = ud_df.merge(song_df, how='left', left_on='song_id', right_on='song_id')

combo_songs = gl.SFrame(new_df)
combo_songs['song_name'] = combo_songs['title'] + ' (' + combo_songs['artist_name'] +')'

combo_songs['user', 'title', 'artist_name', 'plays', 'song_name'].head(4)

Columns:
	user	str
	title	str
	artist_name	str
	plays	int
	song_name	str

Rows: 4

Data:
+-------------------------------+-------------------------------+
|              user             |             title             |
+-------------------------------+-------------------------------+
| fd50c4007b68a3737fe052d5a4... |         You're The One        |
| fd50c4007b68a3737fe052d5a4... | Horn Concerto No. 4 in E f... |
| fd50c4007b68a3737fe052d5a4... |            Tive Sim           |
| fd50c4007b68a3737fe052d5a4... | Catch You Baby (Steve Pitr... |
+-------------------------------+-------------------------------+
+-------------------------------+-------+--------------------------------+
|          artist_name          | plays |           song_name            |
+-------------------------------+-------+--------------------------------+
|         Dwight Yoakam         |   1   | You're The One (Dwight Yoakam) |
| Barry Tuckwell/Academy of ... |   1   | Horn Concerto No. 4 in E f...  |
|            Cartola            |   1   |       Tive Sim (Cartola)       |
|         Lonnie Gordon         |   1   | Catch You Baby (Steve Pitr...  |
+-------------------------------+-------+--------------------------------+
[4 rows x 5 columns]

Collaborative filtering based on song and user is the primary objective.  In addition, the user and artist data will be analyzed in the same fashion to determine the preferred method.   Both models will be trained, tested, and compared.  

In [9]:
df_artist = new_df.groupby(by=['user','artist_name'])[['plays']].sum().reset_index()

artist = gl.SFrame(df_artist)
artist.head(4)

Columns:
	user	str
	artist_name	str
	plays	int

Rows: 4

Data:
+-------------------------------+-------------+-------+
|              user             | artist_name | plays |
+-------------------------------+-------------+-------+
| 00007a02388c208ea7176479f6... |    Dredg    |   3   |
| 00007a02388c208ea7176479f6... |   Local H   |   1   |
| 00007a02388c208ea7176479f6... |     M83     |   2   |
| 00007a02388c208ea7176479f6... |   No Doubt  |   1   |
+-------------------------------+-------------+-------+
[4 rows x 3 columns]

### Data Visualization

The dataset being utilized for the project is significant in size.   Due to the size, the visuals can be challenging to present.  To provide an image of the number of songs and the number of plays, each entry was categorized based on the number of plays.  The pie chart below shows the percentage of songs in each category.   

In [7]:
#Create the category function
def f(row):
    if row['plays'] <= 1:
        val = ' 0-1'
    elif row['plays'] <= 5:
        val = ' 2-5'
    elif row['plays'] <= 10:
        val =  ' 6-10'
    elif row['plays'] <= 50:
        val =  '11-50'
    elif row['plays'] <= 75:
        val =  '51-75'
    elif row['plays'] >75:
        val =  'Greater than 75'
    return val

#Apply the category function 
new_df['category'] = new_df.apply(f, axis=1)

#Confirm the new category added
new_df.head(3)

user             song_id  plays  \
0  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOBONKR12A58A7A7E0      1   
1  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOEGIYH12A6D4FC0E3      1   
2  fd50c4007b68a3737fe052d5a4f78ce8aa117f3d  SOFLJQZ12A6D4FADA6      1   

                                               title  \
0                                     You're The One   
1  Horn Concerto No. 4 in E flat K495: II. Romanc...   
2                                           Tive Sim   

                           release  \
0               If There Was A Way   
1  Mozart - Eine kleine Nachtmusik   
2                 Nova Bis-Cartola   

                                         artist_name  year category  
0                                      Dwight Yoakam  1990      0-1  
1  Barry Tuckwell/Academy of St Martin-in-the-Fie...     0      0-1  
2                                            Cartola  1974      0-1

In [8]:
#Group by categories

pie_chart = new_df.groupby(by=['category'])[['plays']].count().reset_index()

pie_chart.head()

category   plays
0      0-1  862354
1      2-5  451292
2     6-10  100857
3    11-50   72879
4    51-75    2699

In [9]:
#Plot
plt.pie(pie_chart['plays'], labels=pie_chart['category'], autopct='%1.1f%%', shadow=True, startangle=140)
plt.axis('equal')
plt.show()

([<matplotlib.patches.Wedge at 0x12113a910>,
  <matplotlib.text.Text at 0x12119d750>])

In [ ]:
#Apply the category function 
df_artist['category'] = df_artist.apply(f, axis=1)
                                  
art_pc = df_artist.groupby(by=['category'])[['plays']].count().reset_index()
art_pc.head()

plt.pie(art_pc['plays'], labels=art_pc['category'], autopct='%1.1f%%', shadow=True, startangle=140)
plt.axis('equal')
plt.show()

With the song data, over half of the information is based on a song being played a single time.  All data will remain in the analysis; however, the model will be chosen to consider the number of times the song was played.

### GraphLab Canvas

In [23]:
combo_songs.show()

In [24]:
artist.show()

Executing the `graphlab show` commands above will provide summary information about the data in a local browser using GraphLab Canvas.   

The song data includes over 110,000 users listening to over 166,433 unique songs.   The data includes 138,499 unqiue song titles.   The inference is 27,934 songs have the same title although they are different songs.   Analysis will need to be done on the `song_id` rather than the `title`.   Songs were played between 1 times and 923 times.  Both the song data as well as the artist data contain 28,266 unique artists.   Neither dataset has missing values.  The year has some missing values that are populated with 0.  Detail on the missing years can be found in the GraphLab Canvas.   Due to 20% of the year data missing, analysis will not include the year informaiton.  

Since the user identification is based on a user ID, the possiblities of a single person having multiple user IDs or a single user ID being used by multiple individuals cannot be eliminated.   Since the sharing of user IDs is common with families, our analysis will proceed with caution.  

## Train and Adjust parameters

In [12]:
combo_songs.head(3)

Columns:
	user	str
	song_id	str
	plays	int
	title	str
	release	str
	artist_name	str
	year	int
	song_name	str

Rows: 3

Data:
+-------------------------------+--------------------+-------+
|              user             |      song_id       | plays |
+-------------------------------+--------------------+-------+
| fd50c4007b68a3737fe052d5a4... | SOBONKR12A58A7A7E0 |   1   |
| fd50c4007b68a3737fe052d5a4... | SOEGIYH12A6D4FC0E3 |   1   |
| fd50c4007b68a3737fe052d5a4... | SOFLJQZ12A6D4FADA6 |   1   |
+-------------------------------+--------------------+-------+
+-------------------------------+-------------------------------+
|             title             |            release            |
+-------------------------------+-------------------------------+
|         You're The One        |       If There Was A Way      |
| Horn Concerto No. 4 in E f... | Mozart - Eine kleine Nacht... |
|            Tive Sim           |        Nova Bis-Cartola       |
+-------------------------------+-------------------------------+
+-------------------------------+------+--------------------------------+
|          artist_name          | year |           song_name            |
+-------------------------------+------+--------------------------------+
|         Dwight Yoakam         | 1990 | You're The One (Dwight Yoakam) |
| Barry Tuckwell/Academy of ... |  0   | Horn Concerto No. 4 in E f...  |
|            Cartola            | 1974 |       Tive Sim (Cartola)       |
+-------------------------------+------+--------------------------------+
[3 rows x 8 columns]

#### Song Recommendation
The following is the training and testing of the song usage data.  The analysis includes both the user and the song information.    

A recommender based on the plays from users will be created.  The recommender will suggest 5 songs for each user based on the feedback from other users.

In [13]:
train_song, test_song = gl.recommender.util.random_split_by_user(combo_songs,
                                                      user_id="user", item_id="song_id",
                                                      item_test_proportion=0.2)

song_rec = gl.recommender.item_similarity_recommender.create(train_song,
                                                             user_id="user",
                                                             item_id="song_id",
                                                             target="plays",
                                                             only_top_k=5,
                                                             similarity_type="cosine")

rmse_results = song_rec.evaluate(test_song)

print rmse_results.viewkeys()

Recsys training: model = item_similarity

Warning: Ignoring columns title, release, artist_name, year, song_name;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 1489352 observations with 110000 users and 163095 items.

Data prepared in: 1.589s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 3.636ms                        | 0.75       |

| 48.46ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing the 32 most common items by brute force search.

Transposing user-item lists for use in nearest neighbor search.

+------------------------------------+------------+

| Elapsed Time (Data Transposition)) | % Complete |

+------------------------------------+------------+

| 15.682ms                           | 0          |

| 367.691ms                          | 100        |

+------------------------------------+------------+

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 466.835ms                           | 0                | 0               |

+-------------------------------------+------------------+-----------------+

Processing data in one pass using sparse lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.46s                               | 25.5             | 41898           |

| 2.47s                               | 90               | 146947          |

| 3.03s                               | 100              | 163095          |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 4.21949s


Precision and recall summary statistics by cutoff
+--------+------------------+------------------+
| cutoff |  mean_precision  |   mean_recall    |
+--------+------------------+------------------+
|   1    | 0.00463499420626 | 0.00225955967555 |
|   2    | 0.00463499420626 | 0.00414252607184 |
|   3    | 0.00502124372345 | 0.00643243392374 |
|   4    | 0.00463499420626 | 0.0074808254704  |
|   5    | 0.00393974507532 | 0.00786707498758 |
|   6    | 0.00347624565469 | 0.00815676212548 |
|   7    | 0.00430392319152 | 0.0116054185289  |
|   8    | 0.00420046349942 | 0.0124551674668  |
|   9    | 0.00424874468907 |  0.015517574353  |
|   10   | 0.00428736964079 | 0.0161079135876  |
+--------+------------------+------------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 6.278726095375254)

Per User RMSE (best)
+-------------------------------+-------+----------------+
|              user             | count |      rmse      |
+-------------------------------+-------+----------------+
|

In [14]:
rmse_results['rmse_overall']

6.278726095375254

In [16]:
rmse_results['rmse_by_user'].head(4)

Columns:
	user	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------------------------+-------+---------------+
|              user             | count |      rmse     |
+-------------------------------+-------+---------------+
| c174d59c27220c94999bf67cd9... |   4   |      1.0      |
| 1be569aefb976bae9aa6a29af4... |   9   | 4.20317340431 |
| e8f5730a122c64e706a3fb9b47... |   8   | 3.42336455287 |
| 9b9fb6acabe8bdde3012ddc216... |   7   | 1.88982236505 |
+-------------------------------+-------+---------------+
[4 rows x 3 columns]

In [17]:
rmse_results['rmse_by_item'].head(4)

Columns:
	song_id	str
	count	int
	rmse	float

Rows: 4

Data:
+--------------------+-------+------+
|      song_id       | count | rmse |
+--------------------+-------+------+
| SOEKEWC12A8C132A59 |   1   | 1.0  |
| SOQIPYO12AAF3B5B1D |   1   | 10.0 |
| SOYIOZB12A58A797FC |   1   | 4.0  |
| SOSZAST12A6D4F6245 |   1   | 2.0  |
+--------------------+-------+------+
[4 rows x 3 columns]

In [25]:
train_song, test_song = gl.recommender.util.random_split_by_user(combo_songs,
                                                      user_id="user", item_id="song_id",
                                                      item_test_proportion=0.2)

song_rec = gl.recommender.item_similarity_recommender.create(train_song,
                                                             user_id="user",
                                                             item_id="song_id",
                                                             target="plays",
                                                             only_top_k=5,
                                                             similarity_type="pearson")

rmse_results = song_rec.evaluate(test_song)

rmse_results.viewkeys()

Recsys training: model = item_similarity

Warning: Ignoring columns title, release, artist_name, year, song_name;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 1489352 observations with 110000 users and 163095 items.

Data prepared in: 1.38976s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.262ms                        | 0.75       |

| 61.428ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing the 32 most common items by brute force search.

Transposing user-item lists for use in nearest neighbor search.

+------------------------------------+------------+

| Elapsed Time (Data Transposition)) | % Complete |

+------------------------------------+------------+

| 17.843ms                           | 0          |

| 343.022ms                          | 100        |

+------------------------------------+------------+

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 464.289ms                           | 0                | 0               |

+-------------------------------------+------------------+-----------------+

Processing data in one pass using sparse lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.47s                               | 18.75            | 30831           |

| 2.47s                               | 62.75            | 102634          |

| 3.47s                               | 88               | 143701          |

| 4.27s                               | 100              | 163095          |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 4.39432s


Precision and recall summary statistics by cutoff
+--------+----------------+-------------+
| cutoff | mean_precision | mean_recall |
+--------+----------------+-------------+
|   1    |      0.0       |     0.0     |
|   2    |      0.0       |     0.0     |
|   3    |      0.0       |     0.0     |
|   4    |      0.0       |     0.0     |
|   5    |      0.0       |     0.0     |
|   6    |      0.0       |     0.0     |
|   7    |      0.0       |     0.0     |
|   8    |      0.0       |     0.0     |
|   9    |      0.0       |     0.0     |
|   10   |      0.0       |     0.0     |
+--------+----------------+-------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 5.471660027577793)

Per User RMSE (best)
+-------------------------------+-------+------+
|              user             | count | rmse |
+-------------------------------+-------+------+
| bb560bb8852897da731e52705e... |   1   | 0.0  |
+-------------------------------+-------+------+
[1 rows x 3 columns]


Per User

dict_keys(['rmse_by_user', 'precision_recall_overall', 'rmse_by_item', 'precision_recall_by_user', 'rmse_overall'])

In [19]:
rmse_results['rmse_overall']

5.4716600275776

In [21]:
rmse_results['rmse_by_user'].head(4)

Columns:
	user	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------------------------+-------+---------------+
|              user             | count |      rmse     |
+-------------------------------+-------+---------------+
| c174d59c27220c94999bf67cd9... |   4   | 1.63401832999 |
| 1be569aefb976bae9aa6a29af4... |   9   | 2.04651934683 |
| e8f5730a122c64e706a3fb9b47... |   8   | 3.12925062706 |
| 9b9fb6acabe8bdde3012ddc216... |   7   |  1.9644652305 |
+-------------------------------+-------+---------------+
[4 rows x 3 columns]

In [22]:
rmse_results['rmse_by_item'].head(4)

Columns:
	song_id	str
	count	int
	rmse	float

Rows: 4

Data:
+--------------------+-------+----------------+
|      song_id       | count |      rmse      |
+--------------------+-------+----------------+
| SOEKEWC12A8C132A59 |   1   | 2.75925925926  |
| SOQIPYO12AAF3B5B1D |   1   | 6.83146067416  |
| SOYIOZB12A58A797FC |   1   | 1.30303030303  |
| SOSZAST12A6D4F6245 |   1   | 0.516129032258 |
+--------------------+-------+----------------+
[4 rows x 3 columns]

The RMSE (root mean square error) is evaluated for both the cosine and pearson model for the item similarity recommender.  The pearson model is the preferred model with the lower RMSE (5.47 compared to 6.28).

#### Artist Recommendation
Music recommendations may be offered in multiple forms.  In addition to the recommendation of songs, the analysis includes the recommendation of music by artist.  The data was trained as tested the same as the song data to offere comparison of the two collaborative filtering results.   

In [26]:
train_art, test_art = gl.recommender.util.random_split_by_user(artist,
                                                      user_id="user", item_id="artist_name",
                                                      item_test_proportion=0.2)

art_rec = gl.recommender.item_similarity_recommender.create(train_art,
                                                             user_id="user",
                                                             item_id="artist_name",
                                                             target="plays",
                                                             only_top_k=5,
                                                             similarity_type="cosine")

rmse_results = art_rec.evaluate(test_art)

rmse_results.viewkeys()

Recsys training: model = item_similarity

Preparing data set.

Data has 1131349 observations with 109996 users and 28359 items.

Data prepared in: 1.21072s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.496ms                        | 0.75       |

| 33.994ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 2.22s                               | 0                | 0               |

| 3.22s                               | 49.75            | 14175           |

| 8.08s                               | 100              | 28359           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 8.18355s


Precision and recall summary statistics by cutoff
+--------+------------------+------------------+
| cutoff |  mean_precision  |   mean_recall    |
+--------+------------------+------------------+
|   1    | 0.0024154589372  | 0.00144927536232 |
|   2    | 0.0024154589372  | 0.00305958132045 |
|   3    | 0.00201288244767 | 0.00346215780998 |
|   4    | 0.00332125603865 | 0.00678341384863 |
|   5    | 0.00434782608696 | 0.0113123993559  |
|   6    | 0.00402576489533 | 0.0122181964573  |
|   7    | 0.0036231884058  | 0.0126207729469  |
|   8    | 0.00347222222222 | 0.0141304347826  |
|   9    | 0.00322061191626 | 0.0143719806763  |
|   10   | 0.00326086956522 | 0.0163345410628  |
+--------+------------------+------------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 13.99781300253604)

Per User RMSE (best)
+-------------------------------+-------+----------------+
|              user             | count |      rmse      |
+-------------------------------+-------+----------------+
|

dict_keys(['rmse_by_user', 'precision_recall_overall', 'rmse_by_item', 'precision_recall_by_user', 'rmse_overall'])

In [27]:
rmse_results['rmse_overall']

13.99781300253604

In [29]:
rmse_results['rmse_by_user'].head(4)

Columns:
	user	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------------------------+-------+---------------+
|              user             | count |      rmse     |
+-------------------------------+-------+---------------+
| c174d59c27220c94999bf67cd9... |   4   |      1.0      |
| 1be569aefb976bae9aa6a29af4... |   4   | 10.4642247682 |
| cfa3a9fb3d859ace61a1e05fd5... |   4   | 8.09320702812 |
| e8f5730a122c64e706a3fb9b47... |   7   | 9.53904623816 |
+-------------------------------+-------+---------------+
[4 rows x 3 columns]

In [30]:
rmse_results['rmse_by_item'].head(4)

Columns:
	artist_name	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------+-------+--------------+
| artist_name | count |     rmse     |
+-------------+-------+--------------+
|    Spoon    |   2   | 6.0827625303 |
|   Scooter   |   1   |     5.0      |
|     Pru     |   1   |     1.0      |
|  Dragonland |   1   |     1.0      |
+-------------+-------+--------------+
[4 rows x 3 columns]

In [31]:
train_art, test_art = gl.recommender.util.random_split_by_user(artist,
                                                      user_id="user", item_id="artist_name",
                                                      item_test_proportion=0.2)

art_rec = gl.recommender.item_similarity_recommender.create(train_art,
                                                             user_id="user",
                                                             item_id="artist_name",
                                                             target="plays",
                                                             only_top_k=5,
                                                             similarity_type="pearson")

rmse_results = art_rec.evaluate(test_art)

print rmse_results.viewkeys()

Recsys training: model = item_similarity

Preparing data set.

Data has 1131349 observations with 109996 users and 28359 items.

Data prepared in: 1.18551s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.916ms                        | 5.25       |

| 54.265ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.97s                               | 0                | 0               |

| 2.98s                               | 45.5             | 12932           |

| 3.98s                               | 82.25            | 23358           |

| 7.22s                               | 100              | 28359           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 7.30669s


Precision and recall summary statistics by cutoff
+--------+----------------+-------------+
| cutoff | mean_precision | mean_recall |
+--------+----------------+-------------+
|   1    |      0.0       |     0.0     |
|   2    |      0.0       |     0.0     |
|   3    |      0.0       |     0.0     |
|   4    |      0.0       |     0.0     |
|   5    |      0.0       |     0.0     |
|   6    |      0.0       |     0.0     |
|   7    |      0.0       |     0.0     |
|   8    |      0.0       |     0.0     |
|   9    |      0.0       |     0.0     |
|   10   |      0.0       |     0.0     |
+--------+----------------+-------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 13.234525217052376)

Per User RMSE (best)
+-------------------------------+-------+------+
|              user             | count | rmse |
+-------------------------------+-------+------+
| 7fd613423ff2305c23ab30d5f9... |   1   | 0.0  |
+-------------------------------+-------+------+
[1 rows x 3 columns]


Per Use

In [32]:
rmse_results['rmse_overall']

13.234525217052376

In [33]:
rmse_results['rmse_by_user'].head(4)

Columns:
	user	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------------------------+-------+---------------+
|              user             | count |      rmse     |
+-------------------------------+-------+---------------+
| c174d59c27220c94999bf67cd9... |   4   | 2.93748076643 |
| 1be569aefb976bae9aa6a29af4... |   4   | 7.23482442001 |
| cfa3a9fb3d859ace61a1e05fd5... |   4   | 6.13846481143 |
| e8f5730a122c64e706a3fb9b47... |   7   | 7.54462252268 |
+-------------------------------+-------+---------------+
[4 rows x 3 columns]

In [34]:
rmse_results['rmse_by_item'].head(4)

Columns:
	artist_name	str
	count	int
	rmse	float

Rows: 4

Data:
+-------------+-------+----------------+
| artist_name | count |      rmse      |
+-------------+-------+----------------+
|    Spoon    |   2   | 2.17400503742  |
|   Scooter   |   1   | 0.283068783069 |
|     Pru     |   1   |      0.45      |
|  Dragonland |   1   | 0.533333333333 |
+-------------+-------+----------------+
[4 rows x 3 columns]

## Evaluate and Compare

The recommendation models were run for the artist prediction to offer a comparison.   Although it may be expected that the recommendation model would be better with the same amount plays data yet fewer recommendations; however, that was incorrect.  Both models for the artist prediction were significantly higher than the song recommendation counterpart.   

The model for song recommendation using the pearsons similarity type will be used for the remainder of the analysis.  

## Visualize Results

In [35]:
#Preferred method selected above
train_song, test_song = gl.recommender.util.random_split_by_user(combo_songs,
                                                      user_id="user", item_id="song_id",
                                                      item_test_proportion=0.2)

song_rec = gl.recommender.item_similarity_recommender.create(train_song,
                                                             user_id="user",
                                                             item_id="song_id",
                                                             target="plays",
                                                             only_top_k=5,
                                                             similarity_type="pearson")

rmse_results = song_rec.evaluate(test_song)

print rmse_results.viewkeys()

Recsys training: model = item_similarity

Warning: Ignoring columns title, release, artist_name, year, song_name;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 1489352 observations with 110000 users and 163095 items.

Data prepared in: 1.40951s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.854ms                        | 1.75       |

| 68.484ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing the 32 most common items by brute force search.

Transposing user-item lists for use in nearest neighbor search.

+------------------------------------+------------+

| Elapsed Time (Data Transposition)) | % Complete |

+------------------------------------+------------+

| 15.742ms                           | 0          |

| 355.642ms                          | 100        |

+------------------------------------+------------+

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 484.654ms                           | 0                | 0               |

+-------------------------------------+------------------+-----------------+

Processing data in one pass using sparse lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.48s                               | 27.75            | 45456           |

| 2.49s                               | 90               | 146867          |

| 2.95s                               | 100              | 163095          |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 3.04957s


Precision and recall summary statistics by cutoff
+--------+----------------+-------------+
| cutoff | mean_precision | mean_recall |
+--------+----------------+-------------+
|   1    |      0.0       |     0.0     |
|   2    |      0.0       |     0.0     |
|   3    |      0.0       |     0.0     |
|   4    |      0.0       |     0.0     |
|   5    |      0.0       |     0.0     |
|   6    |      0.0       |     0.0     |
|   7    |      0.0       |     0.0     |
|   8    |      0.0       |     0.0     |
|   9    |      0.0       |     0.0     |
|   10   |      0.0       |     0.0     |
+--------+----------------+-------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 5.471660027577793)

Per User RMSE (best)
+-------------------------------+-------+------+
|              user             | count | rmse |
+-------------------------------+-------+------+
| bb560bb8852897da731e52705e... |   1   | 0.0  |
+-------------------------------+-------+------+
[1 rows x 3 columns]


Per User

In [ ]:
from matplotlib import pyplot as plt
%matplotlib inline
plt.style.use('ggplot')

rmsevals = rmse_results['rmse_by_item']['rmse']

plt.hist(rmsevals, bins=20)
plt.title('RMSE by Song')
plt.show()

The RMSE of most songs is relatively low.   However, we can see a few songs with very high RMSE or errors in the prediction from the training set to the testing set.   

# To do:  find way to list highest RMSE songs

In [ ]:
from matplotlib import pyplot as plt
%matplotlib inline
plt.style.use('ggplot')

rmsevals = rmse_results['rmse_by_user']['rmse']
rmsevals = rmsevals[rmse_results['rmse_by_user']['count']>10]


plt.hist(rmsevals, bins=20)
plt.title('RMSE by User')
plt.show()

A few users have high RMSE.   As mentioned in the early analysis of the data, errors for the prediction of certain users may be related to multiple family members accessing the music using the same user ID.  In additon, the high RMSE may be due to eclectic music plays.

### Ranking Factorization Recommender
The ranking factorication recommender learns factors from the user and item.  This model of prediction is preferred for rankings.  Since the plays data is similar to ranking, we will explore the effectiveness of this model.   This model is also know as cross validating collaborative filtering.   The model is very time consuming to process with a large dataset. 

In [36]:
recl = gl.recommender.ranking_factorization_recommender.create(train_song,
                                                             user_id="user",
                                                             item_id="song_id",
                                                             target="plays")

rmse_results = recl.evaluate(test_song)

Recsys training: model = ranking_factorization_recommender

Preparing data set.

Data has 1489352 observations with 110000 users and 163095 items.

Data prepared in: 3.72523s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| ranking_regularization         | Rank-based Regularization Weight                 | 0.25     |

| max_iterations                 | Maximum Number of Iterations                     | 25       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 186169 / 1489352 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 7.14286           | Not Viable                               |

| 1       | 1.78571           | Not Viable                               |

| 2       | 0.446429          | Not Viable                               |

| 3       | 0.111607          | 48.1526                                  |

| 4       | 0.0558036         | 55.4999                                  |

| 5       | 0.0279018         | 61.0443                                  |

| 6       | 0.0139509         | 68.6158                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.111607          | 48.1526                                  |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 124us        | 96.5425           | 7.0175                |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 3.68s        | 121.912           | 9.55683               | 0.111607    |

| 2       | 7.46s        | 100.465           | 8.97077               | 0.111607    |

| 3       | 10.53s       | 103.393           | 9.11631               | 0.111607    |

| 4       | 13.99s       | 118.442           | 9.79481               | 0.111607    |

| 5       | 16.87s       | 150.284           | 11.1075               | 0.111607    |

| 6       | 20.99s       | DIVERGED          | DIVERGED              | 0.111607    |

| RESET   | 22.67s       | 96.5426           | 7.0175                |             |

| 1       | 26.62s       | 92.5759           | 8.12297               | 0.0558036   |

| 2       | 31.16s       | 76.6367           | 7.576                 | 0.0558036   |

| 3       | 35.42s       | 69.2561           | 7.1632                | 0.0558036   |

| 4       | 39.15s       | 65.0597           | 6.92571               | 0.0558036   |

| 5       | 42.21s       | 62.4508           | 6.77912               | 0.0558036   |

| 6       | 45.52s       | 60.5764           | 6.66772               | 0.0558036   |

| 7       | 48.87s       | 58.951            | 6.57274               | 0.0558036   |

| 8       | 52.98s       | 57.6557           | 6.49832               | 0.0558036   |

| 9       | 56.12s       | 56.5821           | 6.43238               | 0.0558036   |

| 10      | 59.10s       | 55.6351           | 6.37449               | 0.0558036   |

| 11      | 1m 2s        | 54.7297           | 6.32029               | 0.0558036   |

| 12      | 1m 5s        | 53.9186           | 6.26998               | 0.0558036   |

| 13      | 1m 8s        | 53.1387           | 6.22326               | 0.0558036   |

| 14      | 1m 11s       | 52.4578           | 6.17832               | 0.0558036   |

| 15      | 1m 16s       | 51.8214           | 6.13845               | 0.0558036   |

| 16      | 1m 19s       | 51.1943           | 6.09854               | 0.0558036   |

| 17      | 1m 23s       | 50.5805           | 6.0599                | 0.0558036   |

| 18      | 1m 27s       | 50.0211           | 6.02356               | 0.0558036   |

| 19      | 1m 30s       | 49.5088           | 5.98812               | 0.0558036   |

| 20      | 1m 34s       | 48.9759           | 5.95359               | 0.0558036   |

| 21      | 1m 37s       | 48.5158           | 5.92135               | 0.0558036   |

| 22      | 1m 40s       | 48.0084           | 5.88865               | 0.0558036   |

| 23      | 1m 42s       | 47.5181           | 5.85669               | 0.0558036   |

| 24      | 1m 45s       | 47.0848           | 5.82534               | 0.0558036   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached (hard limit).

Computing final objective value and training RMSE.

Final objective value: 44.3193

Final training RMSE: 5.61823


Precision and recall summary statistics by cutoff
+--------+-------------------+-------------------+
| cutoff |   mean_precision  |    mean_recall    |
+--------+-------------------+-------------------+
|   1    |  0.00115874855156 | 0.000289687137891 |
|   2    | 0.000579374275782 | 0.000289687137891 |
|   3    | 0.000386249517188 | 0.000289687137891 |
|   4    | 0.000579374275782 | 0.000579374275782 |
|   5    | 0.000463499420626 | 0.000579374275782 |
|   6    | 0.000386249517188 | 0.000579374275782 |
|   7    | 0.000331071014733 | 0.000579374275782 |
|   8    | 0.000289687137891 | 0.000579374275782 |
|   9    | 0.000257499678125 | 0.000579374275782 |
|   10   | 0.000463499420626 | 0.000939873825158 |
+--------+-------------------+-------------------+
[10 rows x 3 columns]

('\nOverall RMSE: ', 5.9122153029011795)

Per User RMSE (best)
+-------------------------------+-------+-----------------+
|              user             | count |       rmse      |
+----------------------------

In [37]:
rmse_results['rmse_overall']

5.9122153029011795

The cross validation model increased the RMSE from 5.47 to 7.34.  

In [38]:
rmse_results['precision_recall_by_user'].groupby('cutoff',[agg.AVG('precision'),agg.STD('precision'),agg.AVG('recall')])

Columns:
	cutoff	int
	Avg of precision	float
	Stdv of precision	float
	Avg of recall	float

Rows: 18

Data:
+--------+-------------------+-------------------+-------------------+
| cutoff |  Avg of precision | Stdv of precision |   Avg of recall   |
+--------+-------------------+-------------------+-------------------+
|   16   | 0.000434530706837 |  0.00519320250348 |  0.00180893523883 |
|   10   | 0.000463499420626 |  0.00679228314705 | 0.000939873825158 |
|   36   | 0.000708124114845 |  0.00457786705229 |  0.0087356765804  |
|   26   | 0.000668508779749 |  0.00502642737494 |  0.00542036822454 |
|   41   | 0.000678291835062 |  0.00417878755405 |  0.0104737994078  |
|   3    | 0.000386249517188 |  0.0113402226774  | 0.000289687137891 |
|   1    |  0.00115874855156 |  0.0340206680322  | 0.000289687137891 |
|   6    | 0.000386249517188 |  0.00801409575946 | 0.000579374275782 |
|   11   |  0.00042136310966 |  0.00617480286095 | 0.000939873825158 |
|   2    | 0.000579374275782 |  0.0170103340161  | 0.000289687137891 |
+--------+-------------------+-------------------+-------------------+
[18 rows x 4 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

The aggregate of the precision and recall are evaluated for this model.   Precision is the measurement of not predicting falsely.  Recall is the measurement of correct predictions.  Accuracy is the measure the fraction of prediting correct.  

### Determine correct number of latent factors

Since the RMSE decreased with the cross validation, the model will be slightly adjusted from the default to analyze again with the number latent factors clearly designated.  Before determining the correct number of latent factors, models will be run for 4 different possible latent factors.  

In [39]:
params = {'user_id': 'user',
         'item_id': 'song_id',
         'target': 'plays',
         'num_factors': [4, 10, 20],
         'regularization': [0.001] ,
         'linear_regularization': [0.001]}

job = gl.model_parameter_search.create( (train_song,test_song),
                                      gl.recommender.ranking_factorization_recommender.create,
                                      params,
                                      max_models=3,
                                      environment=None)


In [ ]:
job.get_status()

In [ ]:
job_result = job.get_results()

job_result.head()

In [ ]:
bst_prms = job.get_best_params()
bst_prms

In [ ]:
models = job.get_models()
models

In [ ]:
comparisonstruct = gl.compare(test_song,models)
gl.show_comparison(comparisonstruct,models)

In [ ]:
models[1]

# Based on the results of the above - manually change the num_factors to that number for the latent factors.

In [ ]:
recl = gl.recommender.ranking_factorization_recommender.create(train_song,
                                                             user_id="user",
                                                             item_id="song_id",
                                                             target="plays",
                                                              num_factors=10, 
                                                              regularization=1e-02,
                                                              linear_regularization = 1e-3)

rmse_results = recl.evaluate(test_song)

In [ ]:
rmse_results['rmse_overall']

In [ ]:
rmse_results['precision_recall_by_user'].groupby('cutoff',[agg.AVG('precision'),agg.STD('precision'),agg.AVG('recall')])

The comparison between the models

In [ ]:
comparisonstruct = gl.compare(test_song, [song_rec, recl])

In [ ]:
comparisonstruct = gl.compare(test_song, [song_rec, recl])

In [ ]:
gl.show_comparison(comparisonstruct,[song_rec, recl])

## Summarize the Ramifications

## Deployment

Turi has in depth documentation on how to deploy a GraphLab model. Specifically, there seems to be three ways to make a deployment:

 - Turi Distributed
 - Turi Predictive Services
 - Simply running `predict` against your chosen model
 
 
[Turi Distributed](https://turi.com/learn/userguide/deployment/pipeline-introduction.html) provides a way to distribute to ec2 or hadoop clusters. Turi Distributed also adds some management tools to help with your production environment.

[Turi Predictive Services](https://turi.com/learn/userguide/deployment/pred-intro.html) offers another service to deploy your models. This way may be considered to be more robust as much of the deployment process and management is completely automated.

If the production environment would be on a local environment, you could just load your model locally and run predictions as needed.


## Exceptional Work

We were really excited to see this project in action. We tried all three different ways to deploy our model. The first two "automated" ways to deploy threw errors when we got to the final steps due to some S3 error.

We attempted to create our own small environment. The working example can be seen below:

In [ ]:
from IPython.display import IFrame
IFrame('http://ec2-52-53-178-233.us-west-1.compute.amazonaws.com:3000/', width=700, height=220)

As you can see, if you select a song title, another song will be recommended based on your choice. This project includes having a python script that polls for a `song_id` as its stdin, which then returns a list of recommendations. A nodejs server sits above the python script to provide a fast, lightweight web server to serve the recommendations. The full source code for this project can be found [here](https://github.com/kjprice/graph-lab-demo).

## References

Reference for the dataset:
@INPROCEEDINGS{Bertin-Mahieux2011,
  author = {Thierry Bertin-Mahieux and Daniel P.W. Ellis and Brian Whitman and Paul Lamere},
  title = {The Million Song Dataset},
  booktitle = {{Proceedings of the 12th International Conference on Music Information
	Retrieval ({ISMIR} 2011)}},
  year = {2011},
  owner = {thierry},
  timestamp = {2010.03.07}
  